src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -277,12 +277,31 @@ class Config {
   }
 };
 
+#ifdef _WINDOWS
+static void map_stack_pages(JavaThread* thread, size_t size, address sp) {
+  address new_sp = sp - size;
+  address watermark = thread->stack_overflow_state()->shadow_zone_growth_watermark();
+
+  if (new_sp < watermark) {
+    size_t page_size = os::vm_page_size();
+    address last_touched_page = watermark - StackOverflow::stack_shadow_zone_size();
+    size_t pages_to_touch = align_up(watermark - new_sp, page_size) / page_size;
+    while (pages_to_touch--) {
+      last_touched_page -= page_size;
+      *last_touched_page = 0;
+    }
+    thread->stack_overflow_state()->set_shadow_zone_growth_watermark(new_sp);
```

Comment:
I'm not familiar with the details of this stack management code and am unclear about the role of the `shadow_zone_growth_watermark` here. The banging code in `os::map_stack_shadow_pages` doesn't access it.

Reply:
The shadow zone growth watermark is just an optimization to avoid banging pages that were already touched. It is set to the highest sp (stack growing up) where we banged already (there is a diagram and more explanations in stackOverflow.hpp). So we don't strictly need it but we would incur in unnecessary overhead without it when the size of the frames freezed in the top stackChunk are a couple of pages in size. By checking this reference first we guarantee that almost all the time we won't have to do anything. I added the update of the watermark in os::map_stack_shadow_pages().